File: ios/RNAppPushUpdate.m
#import "RNAppPushUpdate.h"
#import <React/RCTBridge.h>
#import <React/RCTRootView.h>
#import <objc/runtime.h>

@interface RNAppUpdateButtonHandler : NSObject

- (void)onUpdatePressed;

@end

@implementation RNAppUpdateButtonHandler

- (NSString *)getLibraryDirectory {
  NSArray *paths = NSSearchPathForDirectoriesInDomains(NSLibraryDirectory, NSUserDomainMask, YES);
  return [paths firstObject];
}

- (void)onUpdatePressed {
  UIWindow *window = nil;

  for (UIScene *scene in UIApplication.sharedApplication.connectedScenes) {
      if (scene.activationState == UISceneActivationStateForegroundActive &&
          [scene isKindOfClass:[UIWindowScene class]]) {
          
          UIWindowScene *windowScene = (UIWindowScene *)scene;
          window = windowScene.windows.firstObject;
          break;
      }
  }
  if (!window) return;

  NSFileManager *fileManager = [NSFileManager defaultManager];

  NSString *libraryDirPath = [self getLibraryDirectory];
  NSURL *libraryDir = [NSURL fileURLWithPath:libraryDirPath];
  NSURL *bundleDir = [libraryDir URLByAppendingPathComponent:@"Application Support"];
  if (![fileManager fileExistsAtPath:bundleDir.path]) {
    return;
  }

  NSURL *bundleFile = [bundleDir URLByAppendingPathComponent:@"index.ios.bundle"];
  if (![fileManager fileExistsAtPath:bundleFile.path]) {
    return;
  }

  NSString *jsBundlePath = bundleFile.path;
  RCTBridge *bridge = [[RCTBridge alloc] initWithBundleURL:[NSURL fileURLWithPath:jsBundlePath]
                                            moduleProvider:nil
                                             launchOptions:nil];

  NSString *moduleName = [[NSBundle mainBundle] objectForInfoDictionaryKey:@"CFBundleExecutable"] ?: @"";
  RCTRootView *rootView = [[RCTRootView alloc] initWithBridge:bridge moduleName:moduleName initialProperties:nil];

  UIViewController *vc = [UIViewController new];
  vc.view = rootView;
  window.rootViewController = vc;
  [window makeKeyAndVisible];
}

@end

@interface RNAppPushUpdate ()

@property (nonatomic, strong) RCTBridge *bridge;
@property (nonatomic, assign) BOOL initialized;
@property (nonatomic, strong) NSString *baseUrl;
@property (nonatomic, strong) NSString *productKey;

@end

@implementation RNAppPushUpdate

- (NSString *)getJSBundleFileWithWindow:(UIWindow *)window launchOptions:(NSDictionary *)launchOptions {
    NSString *bundlePath = nil;
    NSString *topic = @"";

    NSURL *bundleUrl = [[NSBundle bundleForClass:[self class]] URLForResource:@"RNAppPushUpdate" withExtension:@"bundle"];
    if (bundleUrl) {
        NSBundle *bundle = [NSBundle bundleWithURL:bundleUrl];
        NSString *path = [bundle pathForResource:@"RNAppPushUpdateInfo" ofType:@"plist"];
        NSDictionary *dict = [NSDictionary dictionaryWithContentsOfFile:path];
        if ([dict isKindOfClass:[NSDictionary class]]) {
            self.baseUrl = dict[@"rn_app_push_update_base_url"] ?: @"";
            self.productKey = dict[@"rn_app_push_update_key"];
            topic = dict[@"rn_app_push_update_fcm_update_topic"] ?: @"";
        }
    }
  
  if (!self.initialized) {
    NSString *bundleDirectory = [[self getLibraryDirectory] stringByAppendingPathComponent:@"Application Support"];
    NSFileManager *fileManager = [NSFileManager defaultManager];
    
    if (![fileManager fileExistsAtPath:bundleDirectory]) {
      NSError *error = nil;
      [fileManager createDirectoryAtPath:bundleDirectory withIntermediateDirectories:YES attributes:nil error:&error];
      if (error) {
        NSLog(@"RNAppPushUpdate ❌ Error creating directory: %@", error.localizedDescription);
      }
    }
    
    NSString *bundleFile = [bundleDirectory stringByAppendingPathComponent:@"index.ios.bundle"];
    if ([fileManager fileExistsAtPath:bundleFile]) {
      NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
      NSString *downloadedVersionCode = [defaults stringForKey:@"rn_app_push_update_shared_prefs_version_code"] ?: @"-1";
      NSString *versionCode = [[NSBundle mainBundle] objectForInfoDictionaryKey:@"CFBundleVersion"] ?: @"0";
      if (downloadedVersionCode == versionCode) {
        NSLog(@"RNAppPushUpdate ✅ Found the latest bundle");
        bundlePath = bundleFile;
      } else {
        NSLog(@"RNAppPushUpdate: Downloaded bundle is for versionCode: %@, current versionCode: %@. Deleting this bundle.", downloadedVersionCode, versionCode);
        
        NSError *error = nil;
        
        if ([fileManager fileExistsAtPath:bundleFile]) {
          BOOL success = [fileManager removeItemAtPath:bundleFile error:&error];
          if (success) {
            NSLog(@"RNAppPushUpdate: ✅ Bundle file deleted.");
            [[NSUserDefaults standardUserDefaults] removeObjectForKey:@"rn_app_push_update_shared_prefs_bundle_id"];
            [[NSUserDefaults standardUserDefaults] removeObjectForKey:@"rn_app_push_update_shared_prefs_version_code"];
          } else {
            NSLog(@"RNAppPushUpdate: ⚠️ Failed to delete bundle file: %@", error.localizedDescription);
          }
        }
      }
    }
  }

    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_BACKGROUND, 0), ^{
        [self checkForUpdate];
    });

    self.initialized = YES;

    if (bundlePath) {
        NSURL *bundleURL = [NSURL fileURLWithPath:bundlePath];
        self.bridge = [[RCTBridge alloc] initWithBundleURL:bundleURL moduleProvider:nil launchOptions:launchOptions];

        NSString *moduleName = [[NSBundle mainBundle] objectForInfoDictionaryKey:@"CFBundleExecutable"] ?: @"";
        RCTRootView *rootView = [[RCTRootView alloc] initWithBridge:self.bridge moduleName:moduleName initialProperties:nil];

        UIViewController *rootVC = [UIViewController new];
        rootVC.view = rootView;
        window.rootViewController = rootVC;
        [window makeKeyAndVisible];
    }

    return bundlePath;
}

- (void)checkForUpdate {
    NSString *key = [[NSBundle mainBundle] objectForInfoDictionaryKey:@"rn_app_push_update_key"] ?: self.productKey ?: @"";
    if ([key isEqualToString:@"no_key"] || [key length] == 0) {
        NSLog(@"RNAppPushUpdate ❌ No key provided in Info.plist. Please refer to the documentation.");
        return;
    }

    NSString *versionCode = [[NSBundle mainBundle] objectForInfoDictionaryKey:@"CFBundleVersion"] ?: @"0";
    NSString *urlStr = [NSString stringWithFormat:@"%@product/versions?key=%@&version_code=%@", self.baseUrl, key, versionCode];
    NSURL *url = [NSURL URLWithString:urlStr];
    if (!url) {
        NSLog(@"RNAppPushUpdate ❌ Invalid URL");
        return;
    }

    NSURLSessionDataTask *task = [[NSURLSession sharedSession] dataTaskWithURL:url completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
        if (error) {
            NSLog(@"RNAppPushUpdate ❌ Error in fetching product versions: %@", error.localizedDescription);
            return;
        }
        if (!data) {
            NSLog(@"RNAppPushUpdate ❌ No data received");
            return;
        }
        NSError *jsonError;
        NSDictionary *json = [NSJSONSerialization JSONObjectWithData:data options:0 error:&jsonError];
        if (jsonError) {
            NSLog(@"RNAppPushUpdate ❌ JSON parse error: %@", jsonError.localizedDescription);
            return;
        }
        BOOL isVersionAccepted = [json[@"is_version_accepted"] boolValue];
        NSString *downloadUrl = json[@"download_url"];
        NSInteger bundleId = [json[@"accepted_bundle_id"] integerValue];
        if (isVersionAccepted && downloadUrl.length > 0) {
            NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
            NSInteger downloadedBundleId = [defaults integerForKey:@"rn_app_push_update_shared_prefs_bundle_id"];
            if (downloadedBundleId != bundleId) {
                [self downloadBundleFromServerWithUrl:downloadUrl bundleId:bundleId versionCode:versionCode];
            } else {
                NSLog(@"RNAppPushUpdate ✅ Latest bundle already downloaded");
            }
        } else {
            NSLog(@"RNAppPushUpdate ⚠️ No download URL or version not accepted");
        }
    }];
    [task resume];
}

- (void)downloadBundleFromServerWithUrl:(NSString *)urlStr bundleId:(NSInteger)bundleId versionCode: (NSString *)versionCode {
    NSURL *url = [NSURL URLWithString:urlStr];
    if (!url || ![urlStr rangeOfString:@"://" options:NSRegularExpressionSearch].length) {
        NSLog(@"RNAppPushUpdate ❌ Invalid download URL");
        return;
    }
    NSURLSessionDataTask *task = [[NSURLSession sharedSession] dataTaskWithURL:url completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
        if (error) {
            NSLog(@"RNAppPushUpdate ❌ Error downloading bundle: %@", error.localizedDescription);
            return;
        }
        if (!data) {
            NSLog(@"RNAppPushUpdate ❌ Empty bundle data");
            return;
        }
        NSHTTPURLResponse *httpResponse = (NSHTTPURLResponse *)response;
        if (httpResponse.statusCode != 200) {
            NSLog(@"RNAppPushUpdate ❌ Invalid HTTP response code");
            return;
        }
        NSString *bundleDirectory = [[self getLibraryDirectory] stringByAppendingPathComponent:@"Application Support"];
        NSString *destinationPath = [bundleDirectory stringByAppendingPathComponent:@"index.ios.bundle"];

        NSError *writeError;
        BOOL success = [data writeToFile:destinationPath options:NSDataWritingAtomic error:&writeError];
        if (!success || writeError) {
            NSLog(@"RNAppPushUpdate ❌ Error saving bundle: %@", writeError.localizedDescription);
            return;
        }

        NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
        [defaults setInteger:bundleId forKey:@"rn_app_push_update_shared_prefs_bundle_id"];
        [defaults setObject:versionCode forKey:@"rn_app_push_update_shared_prefs_version_code"];

        NSLog(@"RNAppPushUpdate ✅ Bundle downloaded to %@", destinationPath);
        [self showUpdateHeader];
    }];
    [task resume];
}

- (NSString *)getLibraryDirectory {
    NSArray *paths = NSSearchPathForDirectoriesInDomains(NSLibraryDirectory, NSUserDomainMask, YES);
    return [paths firstObject];
}

- (void)showUpdateHeader {
  dispatch_async(dispatch_get_main_queue(), ^{
    UIWindow *window = nil;
    for (UIScene *scene in UIApplication.sharedApplication.connectedScenes) {
        if (scene.activationState == UISceneActivationStateForegroundActive &&
            [scene isKindOfClass:[UIWindowScene class]]) {
            
            UIWindowScene *windowScene = (UIWindowScene *)scene;
            window = windowScene.windows.firstObject;
            break;
        }
    }
    if (!window) return;

    UIView *rootView = window.rootViewController.view;
    if (!rootView) return;
    
    RNAppUpdateButtonHandler *handler = [[RNAppUpdateButtonHandler alloc] init];

    UIButton *headerView = [UIButton new];
    headerView.translatesAutoresizingMaskIntoConstraints = NO;
    headerView.backgroundColor = [UIColor colorWithRed:0.98 green:0.75 blue:0 alpha:1];
    [headerView addTarget:handler action:@selector(onUpdatePressed) forControlEvents:UIControlEventTouchUpInside];

    UILabel *label = [UILabel new];
    label.translatesAutoresizingMaskIntoConstraints = NO;
    label.text = @"A new update is available";
    label.textColor = [UIColor blackColor];
    label.textAlignment = NSTextAlignmentLeft;
    label.font = [UIFont boldSystemFontOfSize:14];
    [headerView addSubview:label];

    UIButton *button = [UIButton new];
    button.translatesAutoresizingMaskIntoConstraints = NO;
    [button addTarget:handler action:@selector(onUpdatePressed) forControlEvents:UIControlEventTouchUpInside];
    [headerView addSubview:button];

    UILabel *btnLabel = [UILabel new];
    btnLabel.translatesAutoresizingMaskIntoConstraints = NO;
    btnLabel.font = [UIFont boldSystemFontOfSize:14];
    btnLabel.text = @"Update";
    btnLabel.textColor = [UIColor blueColor];
    [button addSubview:btnLabel];

    [rootView addSubview:headerView];
    
    objc_setAssociatedObject(headerView, "updateHandler", handler, OBJC_ASSOCIATION_RETAIN_NONATOMIC);
    objc_setAssociatedObject(button, "updateHandler", handler, OBJC_ASSOCIATION_RETAIN_NONATOMIC);

    [NSLayoutConstraint activateConstraints:@[
      [headerView.topAnchor constraintEqualToAnchor:rootView.safeAreaLayoutGuide.topAnchor],
      [headerView.leadingAnchor constraintEqualToAnchor:rootView.leadingAnchor],
      [headerView.trailingAnchor constraintEqualToAnchor:rootView.trailingAnchor],
      [headerView.heightAnchor constraintEqualToConstant:60],

      [label.leadingAnchor constraintEqualToAnchor:headerView.leadingAnchor constant:15],
      [label.centerYAnchor constraintEqualToAnchor:headerView.centerYAnchor],

      [btnLabel.centerYAnchor constraintEqualToAnchor:button.centerYAnchor],
      [btnLabel.trailingAnchor constraintEqualToAnchor:button.trailingAnchor constant:-15],

      [button.topAnchor constraintEqualToAnchor:headerView.topAnchor],
      [button.trailingAnchor constraintEqualToAnchor:headerView.trailingAnchor],
      [button.bottomAnchor constraintEqualToAnchor:headerView.bottomAnchor],
      [button.leadingAnchor constraintEqualToAnchor:btnLabel.leadingAnchor constant:-15],
    ]];
  });
}

@end

File: ios/RNAppPushUpdate.swift
import Foundation
// import Firebase
import React

@objc public class RNAppPushUpdate: NSObject {

  private var bridge: RCTBridge?
  private var initialized = false
  private var baseUrl: String = ""
  private var productKey: String?

  public func getJSBundleFile(window: UIWindow, launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> String? {
    var bundlePath: String? = nil
    
    var topic: String = ""
    
    if let bundleUrl = Bundle(for: RNAppPushUpdate.self).url(forResource: "RNAppPushUpdate", withExtension: "bundle"),
       let bundle = Bundle(url: bundleUrl),
       let path = bundle.path(forResource: "RNAppPushUpdateInfo", ofType: "plist"),
       let dict = NSDictionary(contentsOfFile: path) as? [String: Any] {
      baseUrl = dict["rn_app_push_update_base_url"] as? String ?? ""
      productKey = dict["rn_app_push_update_key"] as? String
      topic = dict["rn_app_push_update_fcm_update_topic"] as? String ?? ""
    }
    
    if !initialized {
      let fileManager = FileManager.default
      let bundleDirectory = getLibraryDirectory()
      if !fileManager.fileExists(atPath: bundleDirectory.path) {
        do {
          try fileManager.createDirectory(at: bundleDirectory, withIntermediateDirectories: true, attributes: nil)
        } catch {
          print("RNAppPushUpdate", "❌ Error creating directory: \(error.localizedDescription)")
        }
      }
      let bundleFile = bundleDirectory.appendingPathComponent("index.ios.bundle")
      if FileManager.default.fileExists(atPath: bundleFile.path) {
        var fileSize = 0
        do {
          let resourceValue = try bundleFile.resourceValues(forKeys: [.fileSizeKey])
          fileSize = resourceValue.fileSize ?? 0
        } catch {}
        if fileSize < 50_000 {}
        let defaults = UserDefaults.standard
        let downloadedVersionCode = defaults.string(forKey: "rn_app_push_update_shared_prefs_version_code") ?? "-1"
        let versionCode = Bundle.main.infoDictionary?["CFBundleVersion"] as? String ?? "0"
        if downloadedVersionCode == versionCode {
          print("RNAppPushUpdate", "Found the latest bundle")
          bundlePath = bundleFile.path
        } else {
          print("RNAppPushUpdate", "Downloaded bundle is for versionCode: \(downloadedVersionCode), current versionCode: \(versionCode). Deleting this bundle.")
          do {
            try FileManager.default.removeItem(at: bundleFile)
            print("RNAppPushUpdate", "✅ Bundle file deleted.")
            UserDefaults.standard.removeObject(forKey: "rn_app_push_update_shared_prefs_bundle_id")
            UserDefaults.standard.removeObject(forKey: "rn_app_push_update_shared_prefs_patch_id")
            UserDefaults.standard.removeObject(forKey: "rn_app_push_update_shared_prefs_version_code")
          } catch {
            print("RNAppPushUpdate", "⚠️ Failed to delete bundle file.")
          }
        }
      }

      DispatchQueue.global(qos: .background).async {
        self.checkForUpdate()
      }

    //   if FirebaseApp.app() == nil {
    //       FirebaseApp.configure()
    //       print("RNAppPushUpdate", "Firebase initialized successfully.")
    //   }

      if !topic.isEmpty {
        //   Messaging.messaging().subscribe(toTopic: topic) { error in
        //       if let error = error {
        //           print("Firebase: Failed to subscribe to topic: \(topic)", error)
        //       } else {
        //           print("Firebase: Successfully subscribed to topic: \(topic)")
        //       }
        //   }
      }

      initialized = true
    }
    
    if let jsBundleFilePath = bundlePath {
      bridge = RCTBridge(bundleURL: URL(fileURLWithPath: jsBundleFilePath), moduleProvider: nil, launchOptions: launchOptions)
      let moduleName = Bundle.main.infoDictionary?["CFBundleExecutable"] as? String ?? ""
      let rootView = RCTRootView(bridge: bridge!, moduleName: moduleName, initialProperties: nil)
      window.rootViewController = UIViewController()
      window.rootViewController?.view = rootView
      window.makeKeyAndVisible()
    }
    
    return bundlePath
  }

  private func getBundleFromPrivateDirectory() -> String? {
    let fileManager = FileManager.default
    let bundleDirectory = getLibraryDirectory()
    let filePath = bundleDirectory.appendingPathComponent("index.ios.bundle")
    
    if fileManager.fileExists(atPath: filePath.path) {
      return filePath.path
    } else {
      return nil
    }
  }

  func checkForUpdate() {
//    let baseUrl = Bundle.main.object(forInfoDictionaryKey: "rn_app_push_update_base_url") as? String ?? ""
    let key = Bundle.main.object(forInfoDictionaryKey: "rn_app_push_update_key") as? String ?? productKey ?? ""
    
    if key == "no_key" {
      print("RNAppPushUpdate", "❌ No key provided in Info.plist. Please refer to the documentation.")
      return
    }

    let versionCode = Bundle.main.infoDictionary?["CFBundleVersion"] as? String ?? "0"

    guard let url = URL(string: "\(baseUrl)product/versions?key=\(key)&version_code=\(versionCode)") else {
      print("RNAppPushUpdate", "❌ Invalid URL")
      return
    }

    let client = URLSession.shared
    let request = URLRequest(url: url)
    
    client.dataTask(with: request) { data, response, error in
        if let error = error {
          print("RNAppPushUpdate", "❌ Error in fetching product versions: \(error.localizedDescription)")
          return
        }

        guard let data = data else {
          print("RNAppPushUpdate", "❌ No data received")
          return
        }

        do {
          if let json = try JSONSerialization.jsonObject(with: data, options: []) as? [String: Any] {
            let bundleId = json["bundle_id"] as? Int ?? -1
            let bundleUrl = json["bundle_url"] as? String
            let patchId = json["patch_id"] as? Int ?? -1
            let patchUrl = json["patch_url"] as? String ?? ""
            if let downloadUrl = bundleUrl, !downloadUrl.isEmpty {
              let defaults = UserDefaults.standard
              let downloadedBundleId = defaults.integer(forKey: "rn_app_push_update_shared_prefs_bundle_id")
              let downloadedPatchId = defaults.integer(forKey:
                  "rn_app_push_update_shared_prefs_patch_id")
              
              if downloadedBundleId != bundleId {
                self.downloadBundleFromServer(urlStr: downloadUrl, bundleId: bundleId, patchUrl: patchUrl, patchId: patchId, versionCode: versionCode)
              } else if !patchUrl.isEmpty && downloadedPatchId != patchId {
                self.downloadPatchFromServer(patchUrl: patchUrl, patchId: patchId)
              } else {
                print("RNAppPushUpdate", "Latest bundle already downloaded")
              }
            } else {
              print("RNAppPushUpdate", "No download URL received or version not accepted")
            }
          }
        } catch {
          print("RNAppPushUpdate", "❌ Error in parsing JSON response from server: \(error.localizedDescription)")
        }
    }.resume()
  }

  private func downloadBundleFromServer(urlStr: String, bundleId: Int, patchUrl: String, patchId: Int, versionCode: String) {
    guard let url = URL(string: urlStr), urlStr.range(of: "^[a-zA-Z0-9-]+://", options: .regularExpression) != nil else {
      print("RNAppPushUpdate", "❌ Invalid download URL")
      return
    }

    let client = URLSession.shared
    let request = URLRequest(url: url)

    let task = client.dataTask(with: request) { (data, response, error) in
      if let error = error {
        print("RNAppPushUpdate", "❌ Error in downloading the bundle: \(error.localizedDescription)")
        return
      }

      guard let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 else {
        print("RNAppPushUpdate", "❌ Error in downloading the bundle, unexpected response")
        return
      }

      guard let data = data else {
        print("RNAppPushUpdate", "❌ Received empty or corrupted file, does the bundle exist on the server?")
        return
      }

      do {
        let fileManager = FileManager.default
        let bundleURL = self.getLibraryDirectory()
        let destinationURL = bundleURL.appendingPathComponent("index.ios.bundle")

        try data.write(to: destinationURL)

        let defaults = UserDefaults.standard
        defaults.set(bundleId, forKey: "rn_app_push_update_shared_prefs_bundle_id")
        defaults.set(versionCode, forKey: "rn_app_push_update_shared_prefs_version_code")

        print("RNAppPushUpdate", "✅ File downloaded to \(destinationURL.path)")
        if !patchUrl.isEmpty && patchId > 0 {
          self.downloadPatchFromServer(patchUrl: patchUrl, patchId: patchId)
        } else {
          self.showUpdateHeader()
        }
      } catch {
        print("RNAppPushUpdate", "❌ Error in saving the downloaded bundle: \(error.localizedDescription)")
        let dirPath = self.getLibraryDirectory()
        if FileManager.default.fileExists(atPath: dirPath.path) {
          let bundle = dirPath.appendingPathComponent("index.ios.bundle")
          if FileManager.default.fileExists(atPath: bundle.path) {
            do {
              try FileManager.default.removeItem(at: bundle)
              UserDefaults.standard.removeObject(forKey: "rn_app_push_update_shared_prefs_bundle_id")
              UserDefaults.standard.removeObject(forKey:
                  "rn_app_push_update_shared_prefs_patch_id")
              UserDefaults.standard.removeObject(forKey: "rn_app_push_update_shared_prefs_version_code")
            } catch {}
          }
        }
      }
    }

    task.resume()
  }

  private func downloadPatchFromServer(patchUrl: String, patchId: Int) {
    guard let url = URL(string: patchUrl), patchUrl.range(of: "^[a-zA-Z0-9-]+://", options: .regularExpression) != nil else {
      print("RNAppPushUpdate", "❌ Invalid patch download URL")
      return
    }
    
    let client = URLSession.shared
    let request = URLRequest(url: url)
    
    client.dataTask(with: request) { (data, response, error) in
      if let error = error {
        print("RNAppPushUpdate", "❌ Error in downloading the patch: \(error.localizedDescription)")
        return
      }

      guard let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 else {
        print("RNAppPushUpdate", "❌ Error in downloading the patch, unexpected response")
        return
      }

      guard let data = data else {
        print("RNAppPushUpdate", "❌ Received empty or corrupted patch file, does the patch exist on the server?")
        return
      }
      
      do {
        let bundleURL = self.getLibraryDirectory()
        let destinationURL = bundleURL.appendingPathComponent("bundle.patch")
        
        try data.write(to: destinationURL)
        
        defer {
          do {
            if FileManager.default.fileExists(atPath: destinationURL.path) {
              try FileManager.default.removeItem(at: destinationURL)
            }
          } catch {
            print("RNAppPushUpdate", "Could not delete the patch file, error: \(error.localizedDescription)")
          }
        }

        let defaults = UserDefaults.standard
        defaults.set(patchId, forKey: "rn_app_push_update_shared_prefs_patch_id")
        
        let oldPath = bundleURL.appendingPathComponent("index.android.bundle")
        let patchPath = bundleURL.appendingPathComponent("bundle.patch")
        let outputPath = bundleURL.appendingPathComponent("new_index.android.bundle")
        
        let patchResult = 0
        if patchResult == 0 {
          if FileManager.default.fileExists(atPath: oldPath.path) {
              try FileManager.default.removeItem(at: oldPath)
          }
          try FileManager.default.moveItem(at: outputPath, to: oldPath)
          self.showUpdateHeader()
        } else {
          throw NSError(domain: "RNAppPushUpdate", code: 1, userInfo: [NSLocalizedDescriptionKey: "Apply patch failed"])
        }
      } catch {
        print("RNAppPushUpdate", "❌ Error in parsing the patch downloaded from the server")
        print(error.localizedDescription)
        let bundleURL = self.getLibraryDirectory()
        let patchPath = bundleURL.appendingPathComponent("bundle.patch")
        do {
          if FileManager.default.fileExists(atPath: patchPath.path) {
            try FileManager.default.removeItem(at: patchPath)
            UserDefaults.standard.removeObject(forKey:
                "rn_app_push_update_shared_prefs_patch_id")
          }
        } catch {
          print("RNAppPushUpdate", "Could not delete the patch file, error: \(error.localizedDescription)")
        }
      }
    }.resume()
  }

  private func getLibraryDirectory() -> URL {
    return FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first!
  }
  
  private func showUpdateHeader() {
    DispatchQueue.main.async {
      guard let window = self.getActiveWindow() else { return }
      guard let rootView = window.rootViewController?.view else { return }
      
      let headerView = UIButton()
      headerView.translatesAutoresizingMaskIntoConstraints = false
      headerView.backgroundColor = UIColor(red: 0.98, green: 0.75, blue: 0, alpha: 1)
      headerView.addAction(UIAction { _ in
        self.onUpdatePressed()
      }, for: .touchUpInside)
      
      let label = UILabel()
      label.translatesAutoresizingMaskIntoConstraints = false
      label.text = "A new update is available"
      label.textColor = UIColor(red: 0, green: 0, blue: 0, alpha: 1)
      label.textAlignment = .left
      label.font = UIFont.boldSystemFont(ofSize: 14)
      headerView.addSubview(label)
      
      let button = UIButton()
      button.translatesAutoresizingMaskIntoConstraints = false
      button.addAction(UIAction { _ in
        self.onUpdatePressed()
      }, for: .touchUpInside)
      headerView.addSubview(button)
      
      let btnLabel = UILabel()
      btnLabel.translatesAutoresizingMaskIntoConstraints = false
      btnLabel.font = UIFont.boldSystemFont(ofSize: 14)
      btnLabel.text = "Update"
      btnLabel.textColor = UIColor.blue
      button.addSubview(btnLabel)
      
      rootView.addSubview(headerView)
      
      NSLayoutConstraint.activate([
        headerView.topAnchor.constraint(equalTo: rootView.safeAreaLayoutGuide.topAnchor),
        headerView.leadingAnchor.constraint(equalTo: rootView.leadingAnchor),
        headerView.trailingAnchor.constraint(equalTo: rootView.trailingAnchor),
        headerView.heightAnchor.constraint(equalToConstant: 60),
        label.leadingAnchor.constraint(equalTo: headerView.leadingAnchor, constant: 15),
        label.centerYAnchor.constraint(equalTo: headerView.centerYAnchor),
        btnLabel.centerYAnchor.constraint(equalTo: button.centerYAnchor),
        btnLabel.trailingAnchor.constraint(equalTo: button.trailingAnchor, constant: -15),
        button.topAnchor.constraint(equalTo: headerView.topAnchor),
        button.trailingAnchor.constraint(equalTo: headerView.trailingAnchor),
        button.bottomAnchor.constraint(equalTo: headerView.bottomAnchor),
        button.leadingAnchor.constraint(equalTo: btnLabel.leadingAnchor, constant: -15),
      ])
    }
  }
  
  private func onUpdatePressed() {
    guard let window = getActiveWindow() else { return }

    let fileManager = FileManager.default
    let bundleDirectory = getLibraryDirectory()
    if !fileManager.fileExists(atPath: bundleDirectory.path) {
      return
    }
    let bundleFile = bundleDirectory.appendingPathComponent("index.ios.bundle")
    if !fileManager.fileExists(atPath: bundleFile.path) {
      return
    }
    
    let jsBundlePath = bundleFile.path
    bridge = RCTBridge(bundleURL: URL(fileURLWithPath: jsBundlePath), moduleProvider: nil, launchOptions: nil)
    let moduleName = Bundle.main.infoDictionary?["CFBundleExecutable"] as? String ?? ""
    let rootView = RCTRootView(bridge: bridge!, moduleName: moduleName, initialProperties: nil)

    window.rootViewController = UIViewController()
    window.rootViewController?.view = rootView
    window.makeKeyAndVisible()
  }
  
  func getActiveWindow() -> UIWindow? {
    return UIApplication.shared.connectedScenes
      .filter { $0.activationState == .foregroundActive }
      .compactMap { $0 as? UIWindowScene }
      .flatMap { $0.windows }
      .first
  }
}
